src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -857,61 +907,233 @@ private async Task<Dictionary<KeyValueIdentifier, ConfigurationSetting>> LoadKey
+        private async Task<List<KeyValueChange>> GetRefreshedCollections(
+            IEnumerable<KeyValueSelector> selectors,
+            IDictionary<string, ConfigurationSetting> existingSettings,
+            ConfigurationClient client,
+            CancellationToken cancellationToken)
         {
-            foreach (KeyValueWatcher changeWatcher in _options.MultiKeyWatchers)
+            var keyValueChanges = new List<KeyValueChange>();
+
+            var existingKeys = new HashSet<string>();
+
+            var loadedSettings = new HashSet<string>();
+
+            AsyncPageable<ConfigurationSetting> pageableSettings;
+
+            foreach (KeyValueSelector loadOption in selectors)
             {
-                IEnumerable<ConfigurationSetting> currentKeyValues = GetCurrentKeyValueCollection(changeWatcher.Key, changeWatcher.Label, existingSettings.Values);
+                if (string.IsNullOrEmpty(loadOption.SnapshotName))
+                {
+                    SettingSelector selector = new SettingSelector()
+                    {
+                        KeyFilter = loadOption.KeyFilter,
+                        LabelFilter = loadOption.LabelFilter
+                    };
+
+                    pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+                }
+                else
+                {
+                    ConfigurationSnapshot snapshot;
+
+                    try
+                    {
+                        snapshot = await client.GetSnapshotAsync(loadOption.SnapshotName).ConfigureAwait(false);
+                    }
+                    catch (RequestFailedException rfe) when (rfe.Status == (int)HttpStatusCode.NotFound)
+                    {
+                        throw new InvalidOperationException($"Could not find snapshot with name '{loadOption.SnapshotName}'.", rfe);
+                    }
+
+                    if (snapshot.SnapshotComposition != SnapshotComposition.Key)
+                    {
+                        throw new InvalidOperationException($"{nameof(snapshot.SnapshotComposition)} for the selected snapshot with name '{snapshot.Name}' must be 'key', found '{snapshot.SnapshotComposition}'.");
+                    }
+
+                    pageableSettings = client.GetConfigurationSettingsForSnapshotAsync(
+                        loadOption.SnapshotName,
+                        cancellationToken);
+                }
+
+                await CallWithRequestTracing(async () =>
+                {
+                    await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
+                    {
+                        foreach (ConfigurationSetting setting in page.Values)
+                        {
+                            keyValueChanges.Add(new KeyValueChange()
```

Comment:
In `GetRefreshedCollections` method, we are loading all key-values again but treating them like `KeyValueChange`. My expectation was that any change to watched etags would trigger a full reload of all key-values (like calling LoadSelected again). That saves us from tracking each kv as individually modified or deleted. It also saves us from loading all etags in `UpdateWatchedCollections`. We can break the loop as soon as we see any etag has changed in `UpdateWatchedCollections`.